### PR TITLE
Prototype for m2e-pde source locator

### PR DESCRIPTION
**Pull request by laeubi**

### CODE REVIEW

**Comment by laeubi** [comment]
Still blocked by
- https://github.com/eclipse-pde/eclipse.pde/pull/214

File: org.eclipse.m2e.pde.target/src/org/eclipse/m2e/pde/target/MavenPluginSourcePathLocator.java
Comment (HannesWell) [inline org.eclipse.m2e.pde.target/src/org/eclipse/m2e/pde/target/MavenPluginSourcePathLocator.java:117]: Iterating over all bundles in the TP could be a relatively slow operation if there are many.
It would be ideal if there would be some mean to obtain the TargetBundle of a external Plugin model, but I didn't find any. So maybe it would be reasonable to maintain a corresponding Mapping within M2E that is updated when the TP changes.

**Review by HannesWell — CHANGES_REQUESTED** [review]
In general it works but needs some adjustments to work with artifacts with inherited groupId+ version.
And we should have a look at some potential/assumed performance issues.

**Comment by HannesWell** [comment]
This change should use the `MavenArtifactIdentifier.identify()` and `MavenArtifactIdentifier.resolveSourceLocation()` methods.

**Comment by HannesWell** [comment]
Eclipse 4.26 is now available in the TP.

File: org.eclipse.m2e.pde.target/src/org/eclipse/m2e/pde/target/MavenPluginSourcePathLocator.java
Comment (laeubi) [reply org.eclipse.m2e.pde.target/src/org/eclipse/m2e/pde/target/MavenPluginSourcePathLocator.java:79]: It all depends how you define "faster"... reading the xml might take a few milliseconds more but this actually don't matter much here...

File: org.eclipse.m2e.pde.target/src/org/eclipse/m2e/pde/target/MavenPluginSourcePathLocator.java
Comment (laeubi) [reply org.eclipse.m2e.pde.target/src/org/eclipse/m2e/pde/target/MavenPluginSourcePathLocator.java:99]: There is no way to read the parent from an embedded pom, such poms should be "flatten" already.

**Comment by laeubi** [comment]
> Eclipse 4.26 is now available in the TP.

I updated the PR and added a reverse file lookup, that should be "fast enough"... especially as in my testing in almost all cases the sources are already found beforehands.

**Comment by laeubi** [comment]
> This change should use the MavenArtifactIdentifier.identify() and MavenArtifactIdentifier.resolveSourceLocation() methods.

Done.

**Comment by HannesWell** [comment]
Looks good.

> > Eclipse 4.26 is now available in the TP.
> 
> I updated the PR and added a reverse file lookup, that should be "fast enough"... especially as in my testing in almost all cases the sources are already found beforehands.

Are the sources also resolved during the target-resolution? That would maybe explain it.
However in cases where a artifact from Maven-Central is re-distributed via a P2-repo that does not contain the sources, this is still useful.

**Comment by laeubi** [comment]
Depends on

1. include source enabled in the target
2. one has already run a maven build that uses the artifact and sources are downloaded

**Pull request merged**
